src/main/java/de/epiceric/shopchest/utils/ShopUtils.java

```diff
@@ -278,120 +288,128 @@ public void updateShops(Player player, boolean force) {
+        double itemDistSquared = Math.pow(plugin.getShopChestConfig().maximal_item_distance, 2);
+        double hologramDistSquared = Math.pow(plugin.getShopChestConfig().maximal_distance, 2);
 
-        double i = 0;
+        boolean firstShopInSight = plugin.getShopChestConfig().only_show_first_shop_in_sight;
 
-        do {
-            Location below = loc.clone().subtract(0, 1, 0);
+        // used if only_show_first_shop_in_sight
+        List<Shop> otherShopsInSight = firstShopInSight ? new ArrayList<Shop>() : Collections.<Shop>emptyList();
+        double nearestDistance = 0;
+        Shop nearestShop = null;
 
-            Shop shop = getShop(loc);
-            if (shop != null) {
-                shops.add(shop);
-            } else if ((shop = getShop(below)) != null) {
-                shops.add(shop);
-            }
+        Location pLoc = player.getEyeLocation();
+        Vector pVec = pLoc.toVector();
+        Vector pDir = pLoc.getDirection();
 
-            loc.add(direction);
-            i++;
-        } while (i <= dist - (dist%1));
+        for (Shop shop : getShops()) {
+            Location shopLocation = shop.getLocation();
 
-        direction.multiply(dist - (dist%1));
-        loc.add(direction);
+            if (shopLocation.getWorld().getName().equals(player.getWorld().getName())) {
+                double distanceSquared = shop.getLocation().distanceSquared(player.getLocation());
 
-        Location below = loc.clone().subtract(0, 1, 0);
+                // Display item based on distance
+                if (shop.hasItem()) {
+                    if (distanceSquared <= itemDistSquared) {
+                        shop.getItem().showPlayer(player);
+                    } else {
+                        shop.getItem().hidePlayer(player);
+                    }
+                }
 
-        Shop shop = getShop(loc);
-        if (shop != null) {
-            shops.add(shop);
-        } else if ((shop = getShop(below)) != null) {
-            shops.add(shop);
-        }
+                // Display hologram based on sight
+                if (shop.hasHologram()) {
+                    if (distanceSquared < hologramDistSquared) {
 
-        return shops;
-    }
+                        Location targetLocation = shop.getHologram().getLocation();
+                        targetLocation.setY(shop.getLocation().getY() + 1);
 
-    /**
-     * Update hologram and item of the shop for a player based on their distance to each other
-     * @param shop Shop to update
-     * @param player Player to show the update
-     */
-    public void updateShop(Shop shop, Player player) {
-        double holoDistSqr = Math.pow(plugin.getShopChestConfig().maximal_distance, 2);
-        double itemDistSqr = Math.pow(plugin.getShopChestConfig().maximal_item_distance, 2);
+                        double angle = targetLocation.toVector().subtract(pVec).angle(pDir);
+                        double distance = Math.sqrt(distanceSquared);
```

Comment:
Square root isn't that expensive, I tested and optimised (you'll see in new commit) and got these results (after ±15 minutes of update at `FASTEST` rate):

```
[20:22:12 INFO]: vectorMethod took in average 3207.4862432501927ns
[20:22:12 INFO]: locationMethod took in average 7770.374328472755ns
```

So it's roughly 2x faster.